### DISCUSSION TITLE
Extra layer when running two read/writes within one pipeline

### DISCUSSION THREAD

**Discussion by lewi0622** [abey79/vpype#695] [Q&A] [answered]
File for reference and reproduction
![Demo1](https://github.com/abey79/vpype/assets/61953877/626c60ce-b0a5-49d7-8c65-c985dda83606)

# Single pipeline
```bash
vpype \
read -a id 'C://Users//lewi0//Downloads//Demo1.svg'  occult  -i  write 'C://Users//lewi0//Downloads//Demo1_PROCESSED.svg' \
read -a stroke 'C://Users//lewi0//Downloads//Demo1_PROCESSED.svg'  \
eval "k=0" \
forlayer write "C://Users//lewi0//Downloads//Demo1_PROCESSED"%_i%".svg" end
```
Produces 4 files when it should produce 3:
1. Demo1_PROCESSED.svg
	1. Occulted version of the svg, with two colors and three layers
2. Demo1_PROCESSED0.svg
	1. Just the green layers
3. Demo1_PROCESSED1.svg
	1. Just the red layer
4. Demo1_PROCESSED2.svg
	1. The topmost green layer again for some reason

# Two separate pipelines
```bash
vpype \
read -a id 'C://Users//lewi0//Downloads//Demo1.svg'  occult  -i  write 'C://Users//lewi0//Downloads//Demo1_PROCESSED.svg'
```
```bash
vpype \
read -a stroke 'C://Users//lewi0//Downloads//Demo1_PROCESSED.svg'  \
eval "k=0" \
forlayer write "C://Users//lewi0//Downloads//Demo1_PROCESSED"%_i%".svg" end
```
Produces 3 files correctly

**Comment by abey79** [comment, marked as answer]
This is working as intended.

The `write` command does not flush the pipeline, so the result of the initial `read` + `occult` (ie 3 layers) is still in the pipeline by the time you reach `forlayer`. The second read just adds (redundant) lines in layer 1 and 2. You can see that for yourself by inserting a `show` command:

```
vpype \
read -a id 'C://Users//lewi0//Downloads//Demo1.svg'  occult  -i  write 'C://Users//lewi0//Downloads//Demo1_PROCESSED.svg' \
show \
read -a stroke 'C://Users//lewi0//Downloads//Demo1_PROCESSED.svg'  \
eval "k=0" \
forlayer write "C://Users//lewi0//Downloads//Demo1_PROCESSED"%_i%".svg" end
```

It's possible to run the pipeline in a single line, but it requires the knowledge of which layers to merge:
```
vpype \
read -a id 'C://Users//lewi0//Downloads//Demo1.svg'  occult  -i  \
lmove 3 1  \
eval "k=0" \
forlayer write "C://Users//lewi0//Downloads//Demo1_PROCESSED"%_i%".svg" end
```

It's possible to automatically group layers by color, but the related pipeline is rather complicated IIRC (it's somewhere in the cookbook).

If you want to avoid the intermediate file, you could pipe the two command together (if that works on on windows):
```
vpype \
read -a id 'C://Users//lewi0//Downloads//Demo1.svg'  occult  -i  write - | vpype read -a stroke -  \
eval "k=0" \
forlayer write "C://Users//lewi0//Downloads//Demo1_PROCESSED"%_i%".svg" end
```

Note the use of `-` instead of a file name. In that case, standard in and out are used.
